atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/AbortedTransactionSoftCache.java

```diff
@@ -0,0 +1,166 @@
+            latestTsSeenSoFar = Math.max(latestTsSeenSoFar, maybeSnapshot.get().lastKnownConcludedTimestamp);
+        }
+
+        long latestBucket = Utils.getBucket(latestTsSeenSoFar);
+        // The purpose of this call is to refresh the knownConcluded store for current bucket if it is not up-to-date.
+        // Do not remove this line without considering perf implications.
+        knownConcludedTransactions.isKnownConcluded(latestBucket, KnownConcludedTransactions.Consistency.REMOTE_READ);
+
+        latestTsSeenSoFar = Math.max(latestTsSeenSoFar, knownConcludedTransactions.lastKnownConcludedTimestamp());
+        return latestTsSeenSoFar;
+    }
+
+    private PatchyCache refreshPatchyCache(
+            Optional<PatchyCache> maybeSnapshot, long latestTsSeenSoFar, long latestBucketSeenSoFar) {
+        PatchyCache refreshedPatchyCache;
+
+        if (maybeSnapshot.isEmpty() || maybeSnapshot.get().bucket < latestBucketSeenSoFar) {
+            refreshedPatchyCache = loadPatchyBucket(latestTsSeenSoFar, latestBucketSeenSoFar);
+        } else {
+            // need to load a range of timestamps
+            refreshedPatchyCache = extendPatchIfNeeded(maybeSnapshot.get(), latestTsSeenSoFar);
+        }
+
+        patchyCache = refreshedPatchyCache;
+        return refreshedPatchyCache;
+    }
+
+    @Override
+    public void close() {
+        autobatcher.close();
+    }
+
+    private PatchyCache extendPatchIfNeeded(PatchyCache snapshot, long latestTs) {
+        long currentLastKnownConcluded = snapshot.lastKnownConcludedTimestamp;
+
+        // It is possible that cache was updated by the time we reach here and refresh is not required any more.
+        if (latestTs <= currentLastKnownConcluded) {
+            return snapshot;
+        }
+
+        Set<Long> newAbortedTransactions =
+                futileTimestampStore.getAbortedTransactionsInRange(currentLastKnownConcluded, latestTs);
+        snapshot.extend(latestTs, newAbortedTransactions);
+        return snapshot;
+    }
+
+    private PatchyCache loadPatchyBucket(long latestTsSeenSoFar, long latestBucketSeenSoFar) {
+        long maxTsInCurrentBucket = Utils.getMaxTsInCurrentBucket(latestBucketSeenSoFar);
+
+        Set<Long> futileTimestamps = futileTimestampStore.getAbortedTransactionsInRange(
+                Utils.getMinTsInBucket(latestBucketSeenSoFar), Utils.getMaxTsInCurrentBucket(latestBucketSeenSoFar));
+
+        return new PatchyCache(Math.min(latestTsSeenSoFar, maxTsInCurrentBucket), futileTimestamps);
+    }
+
+    private Optional<PatchyCache> getSnapshot() {
+        return Optional.ofNullable(patchyCache);
+    }
+
+    private TransactionSoftCacheStatus getStatus(long startTimestamp, Set<Long> abortedTransactions) {
+        return abortedTransactions.contains(startTimestamp)
+                ? TransactionSoftCacheStatus.IS_ABORTED
+                : TransactionSoftCacheStatus.IS_NOT_ABORTED;
+    }
+
+    static class PatchyCache {
```

Comment:
May want some comments/docs for reference, but if I'm reading this correctly:
- this represents a partial cache of aborted transactions for the "current" bucket (question: what happens if sweep is more than 1MM timestamps behind?)
- for values in this bucket before or equal to `lastKnownConcludedTimestamp`, `abortedTransactions` is definitive: something in `abortedTransactions` is aborted, but something that is not is committed.
- for values in this bucket after `lastKnownConcludedTimestamp`, `abortedTransactions` identifies transactions that are aborted, but things that are not in the set are unknown.